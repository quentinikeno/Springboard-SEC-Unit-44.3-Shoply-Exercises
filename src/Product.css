.Product .image img {
	width: fit-content;
}
