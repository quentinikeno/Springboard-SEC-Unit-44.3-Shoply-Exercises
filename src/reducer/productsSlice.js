import { createSlice } from "@reduxjs/toolkit";
import inventory from "../data.json";

const initialState = {
	inventory,
	cart: {},
};

export const productsSlice = createSlice({
	name: "products",
	initialState,
	reducers: {
		addToCart: (state, action) => {
			const { id, qty } = action.payload;
			const foundProduct = state.cart[id];
			if (foundProduct) {
				state["cart"][id]["quantity"] += qty;
			} else {
				state["cart"][id] = { quantity: qty };
			}
		},
		removeFromCart: (state, action) => {
			const { id, qty } = action.payload;
			const foundProduct = state.cart[id];
			if (foundProduct) {
				state["cart"][id]["quantity"] -= qty;
				if (state["cart"][id]["quantity"] < 1) delete state["cart"][id];
			} else {
				console.error("Item not found in cart.");
			}
		},
		clearCart: (state) => {
			state.cart = {};
		},
	},
});

export const { addToCart, removeFromCart, clearCart } = productsSlice.actions;

export default productsSlice.reducer;
