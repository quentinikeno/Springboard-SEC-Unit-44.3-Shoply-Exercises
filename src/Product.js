import { useSelector, useDispatch } from "react-redux";
import { addToCart, removeFromCart } from "./reducer/productsSlice";
import "./Product.css";

const Product = ({ id, product }) => {
	const { name, price, image_url } = product;
	const cart = useSelector((state) => state.products.cart);
	const inCart = Object.keys(cart).includes(id);
	const dispatch = useDispatch();

	const handleAddToCart = () => {
		dispatch(addToCart({ id, qty: 1 }));
	};

	const handleRemoveFromCart = () => {
		dispatch(removeFromCart({ id, qty: 1 }));
	};

	return (
		<div className="Product column is-half-desktop is-one-third-widescreen is-one-quarter-fullhd">
			<div className="card">
				<div className="card-image">
					<figure className="image is-4by3">
						<img src={image_url} alt={name} className="mx-auto" />
					</figure>
				</div>
				<div className="card-content">
					<div className="media">
						<div className="media-content">
							<p className="title is-4 is-capitalized">{name}</p>
							<p className="subtitle is-5">${price}</p>
						</div>
					</div>
				</div>
				<footer className="card-footer">
					<div className="card-footer-item">
						<button
							className="button is-primary is-rounded is-outlined is-fullwidth"
							onClick={handleAddToCart}
						>
							Add <i className="fa-solid fa-cart-plus ml-2"></i>
						</button>
					</div>
					{inCart && (
						<>
							<div className="card-footer-item">
								{cart[id]["quantity"]}
							</div>
							<div className="card-footer-item">
								<button
									className="button is-danger is-rounded is-outlined is-fullwidth"
									onClick={handleRemoveFromCart}
								>
									Remove{" "}
									<i className="fa-solid fa-cart-shopping ml-2"></i>
								</button>
							</div>
						</>
					)}
				</footer>
			</div>
		</div>
	);
};

export default Product;
