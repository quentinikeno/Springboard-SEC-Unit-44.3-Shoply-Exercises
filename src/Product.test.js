import { render } from "@testing-library/react";
import { store } from "./reducer/store";
import { Provider } from "react-redux";
import Product from "./Product";

const tvId = "47314fa1-ae56-4eae-80be-af6691145951";
const tv = {
	name: "tv",
	price: 219.99,
	description:
		"A beautiful, big-screen TV. Because hey, Netflix isn't going to watch itself.",
	image_url:
		"https://images.samsung.com/is/image/samsung/latin_en-hd-j4300ah-un32j4300ahxpa-001-front-indigo-blue",
};

it("renders without crashing", () => {
	render(
		<Provider store={store}>
			<Product id={tvId} product={tv} />
		</Provider>
	);
});

it("renders matches snapshot", () => {
	const { asFragment } = render(
		<Provider store={store}>
			<Product id={tvId} product={tv} />
		</Provider>
	);
	expect(asFragment()).toMatchSnapshot();
});
