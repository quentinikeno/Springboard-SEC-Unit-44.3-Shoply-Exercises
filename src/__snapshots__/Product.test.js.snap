// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Product column is-half-desktop is-one-third-widescreen is-one-quarter-fullhd"
  >
    <div
      class="card"
    >
      <div
        class="card-image"
      >
        <figure
          class="image is-4by3"
        >
          <img
            alt="tv"
            class="mx-auto"
            src="https://images.samsung.com/is/image/samsung/latin_en-hd-j4300ah-un32j4300ahxpa-001-front-indigo-blue"
          />
        </figure>
      </div>
      <div
        class="card-content"
      >
        <div
          class="media"
        >
          <div
            class="media-content"
          >
            <p
              class="title is-4 is-capitalized"
            >
              tv
            </p>
            <p
              class="subtitle is-5"
            >
              $219.99
            </p>
          </div>
        </div>
      </div>
      <footer
        class="card-footer"
      >
        <div
          class="card-footer-item"
        >
          <button
            class="button is-primary is-rounded is-outlined is-fullwidth"
          >
            Add 
            <i
              class="fa-solid fa-cart-plus ml-2"
            />
          </button>
        </div>
      </footer>
    </div>
  </div>
</DocumentFragment>
`;
